### PR TITLE
Add WPTs for parse error handling in SharedWorkers

### PR DESCRIPTION
**Pull request by elkurin**
Corresponds to [#5347](https://github.com/whatwg/html/pull/5347)

This pull request adds web-platform-tests to check if parse error events invoked by shared workers are successfully caught from the constructor.

### CODE REVIEW

**Comment by annevk** [comment]
I think we should have tests for multiple `SharedWorker` instances as well.

**Comment by elkurin** [comment]
As we discussed in [#5347](https://github.com/whatwg/html/issues/5323), the behavior of shared workers (not only about error handling) when the one shared worker is constructed several times simultaneously is not defined correctly in the spec because of the raciness you mentioned, so I'm not sure about adding the test cases with multiple SharedWorker instances.

Or did you mean differently? If so, I would appreciate if you give me a brief example.

File: workers/shared-worker-parse-error-failure.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+promise_test(async () => {
+  const scriptURL = 'modules/resources/syntax-error.js';
+  const worker = new SharedWorker(scriptURL, { name: 'classic',
+                                               type: 'classic' });
+  return new Promise(resolve => worker.onerror = resolve);
+}, 'Classic shared worker construction for script with syntax error should ' +
+   'dispatch an ErrorEvent.');
+
+promise_test(async () => {
```
Comment (domenic) [inline workers/shared-worker-parse-error-failure.html]: These tests pass in Firefox because the `import` is a syntax error. Maybe we should add something like this? (I haven't tested this code so it might be wrong.)

```js
// Can put at the top of the test file I think
let supportsModuleWorkers = false;
const options = {
  get type() {
    supportsModuleWorkers = true;
  }
};
new Worker("data:,", options).terminate();

// in each test:
assert_precondition(supportsModuleWorkers, "Supports module workers");
```

File: workers/shared-worker-parse-error-failure.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE html>
+<title>SharedWorker: parse error failure</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script>
+
+promise_test(async () => {
+  const scriptURL = 'modules/resources/syntax-error.js';
+  const worker = new SharedWorker(scriptURL, { name: 'classic',
```
Comment (domenic) [inline workers/shared-worker-parse-error-failure.html]: It's not clear why some of these tests have `name` values and some don't. If it's for debugging I'd expect distinct names on all of them. Otherwise I'd just omit it.

File: workers/shared-worker-parse-error-failure.html [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                                               type: 'classic' });
+  return new Promise(resolve => worker.onerror = resolve);
```
Comment (domenic) [inline workers/shared-worker-parse-error-failure.html]: It would be good to test that the value passed to onerror is an ErrorEvent. Also, worth testing that there are no other arguments, as `window.onerror` gets extra arguments but `worker.onerror` should not.

Maybe that is scope creep for this test... But if we don't want to test those things, then we should change the test description to say `error event` instead of `ErrorEvent`.

File: workers/shared-worker-parse-error-failure.html (hunk 2 above)
Comment (elkurin) [reply workers/shared-worker-parse-error-failure.html]: This is because SharedWorker instances with same name, script url and constructor origin are connected to the same WorkerGlobalScope. As for the first 2 tests, SharedWorker instances have the same script url and constructor origin, so to distinguish them we need to assign the different names here. In this test, it is expected to fail constructing WorkerGlobalScope, so using the same name is not the problem, but it's not nice when the test doesn't run as expected.

Comment (domenic) [reply workers/shared-worker-parse-error-failure.html]: I see. Maybe just give them all unique names then, even if it's not strictly necessary.

File: workers/shared-worker-parse-error-failure.html (hunk 1 above)
Comment (elkurin) [reply workers/shared-worker-parse-error-failure.html]: Thanks for the advice. It's true that we should check if module shared worker is supported, but checking it from if type() is called seems to be not appropriate. In FireFox, module is not supported but WorkerOptions is supported so type() could be called depending on their implementation.

In this test, we only use non-nested static import as a module feature so we can assume module is supported when non-nested static import runs successfully. To check this, the test might timeout when the vendor doesn't support parse error handling, so I think it's better to divide this test into the test for classic and the test for module so that the test for classic runs appropriately even if the test for module timeouts.

File: workers/shared-worker-parse-error-failure.html (hunk 2 above)
Comment (elkurin) [reply workers/shared-worker-parse-error-failure.html]: After this comment, I decided to separate the test into 'classic' and 'module', so I removed the worker's name anyway.

File: workers/shared-worker-parse-error-failure.html (hunk 3 above)
Comment (elkurin) [reply workers/shared-worker-parse-error-failure.html]: I think the event we get here is not defined as ErrorEvent in the spec but Event or ErrorEvent.
https://html.spec.whatwg.org/multipage/indices.html#event-error
I seems better to modify the description as you said.

Comment (elkurin) [reply workers/shared-worker-parse-error-failure.html]: FYI: FireFox returns Event here.
https://bugzilla.mozilla.org/show_bug.cgi?id=1610438

**Comment by annevk** [comment]
I guess I'm okay splitting things up.

File: workers/shared-worker-parse-error-failure.html (hunk 3 above)
Comment (domenic) [reply workers/shared-worker-parse-error-failure.html]: Event is correct. The spec says "fire an event" with no constructor arg, and step 1 of https://dom.spec.whatwg.org/#concept-event-fire says to use `Event`.

The index is a non-normative summary of every event named `error` in the entire HTML spec. You have to look at individual "fire an event" sites to see which constructor they use.

File: workers/modules/shared-worker-parse-error-failure.html [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+let supportsModuleWorkers = false;
+promise_test(async () => {
+  const worker = new SharedWorker('resources/static-import-worker.js',
+                                  { type: 'module' });
+  const msg = await new Promise((resolve, reject) => {
+    worker.port.onmessage = resolve;
+    worker.onerror = reject;
+  });
+  assert_array_equals(msg.data, ['export-on-load-script.js']);
+  if (msg.data.length == 1 && msg.data[0] === 'export-on-load-script.js') {
```
Comment (domenic) [inline workers/modules/shared-worker-parse-error-failure.html]: If the assert fails then this line will not be reached, so this if condition is redundant.

I think the best way to do this is to convert this all into a `promise_setup()`.

```js
promise_setup(async () => {
  const worker = new SharedWorker('resources/static-import-worker.js',
                                  { type: 'module' });

  const supportsModuleWorkers = await new Promise((resolve, reject) => {
    worker.port.onmessage = e => resolve(e.data.length === 1 && e.data[0] == 'export-on-load-script.js');
    worker.onerror = () => resolve(false);
  });

  assert_precondition(supportsModuleWorkers, "Static import must be supported on module shared worker to run this test.");
});
```

Then you can remove the `assert_precondition` from each individual test. Also it makes it so that there are only 2 tests, both with precondition failures, instead of 1 test with a failure and 2 tests with precondition failures.

File: workers/shared-worker-parse-error-failure.html (hunk 3 above)
Comment (domenic) [reply workers/shared-worker-parse-error-failure.html]: Upon reviewing the spec PR I think it's actually important that we test the type (and arguments) here. This is test of new functionality and it's easy to get it wrong. We should be sure that it gets test coverage.

File: workers/modules/shared-worker-parse-error-failure.html (hunk 4 above)
Comment (elkurin) [reply workers/modules/shared-worker-parse-error-failure.html]: Thanks. I modified in this way.

File: workers/shared-worker-parse-error-failure.html (hunk 3 above)
Comment (elkurin) [reply workers/shared-worker-parse-error-failure.html]: I added the test to check if the event's type is 'error'.

Comment (domenic) [reply workers/shared-worker-parse-error-failure.html]: Sorry, by "type" I actually meant "constructor". Please test its `.constructor` property, and also check the number of arguments sent to `onerror`.

Comment (elkurin) [reply workers/shared-worker-parse-error-failure.html]: Sorry for taking it in the wrong way. I modified as you suggested.
I found the test to check onerror for dedicated workers:
https://github.com/web-platform-tests/wpt/blob/master/workers/constructors/Worker/AbstractWorker.onerror.html
and it seems that this test should be also merged into wpt/workers/modules/dedicated-worker-parse-error-failure.html so that we can test parse error occurred in statically imported scripts (but not in this CL).

**Review by bakulf — APPROVED** [review]

**Comment by elkurin** [comment]
@domenic Sorry, I did something wrong with with GIt commits, and got many non-related labels. Could remove those?

File: workers/modules/shared-worker-parse-error-failure.html [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  );
+});
+
+const checkArguments = (args) => {
+  assert_equals(args.a.constructor, Event);
+  assert_true('message' in args.a, 'Event.message');
+  assert_equals(typeof args.a.message, 'string', 'Event.message');
+  assert_equals(args.a.filename, location.href + '/resources/syntax-error.js'),
+                'Event.filename');
+  assert_equals(args.a.lineno, 1, 'Event.lineno');
```
Comment (domenic) [inline workers/modules/shared-worker-parse-error-failure.html]: This doesn't seem right. `Event` does not have `messsage`, `filename`, or `lineno` properties. Does this test somehow pass in Chrome?

I'd expectt the test to check that these properties do _not_ exist...

File: workers/modules/shared-worker-parse-error-failure.html [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  assert_equals(args.b, undefined, 'unexpected second argument to onerror');
+  assert_equals(args.c, undefined, 'unexpected third argument to onerror');
+};
+
+promise_test(async () => {
+  const scriptURL = 'resources/syntax-error.js';
+  const worker = new SharedWorker(scriptURL, { type: 'module' });
+  const args = await new Promise(resolve =>
+      worker.onerror = (a, b, c) => resolve({a, b, c}));
```
Comment (domenic) [inline workers/modules/shared-worker-parse-error-failure.html]: This could be a little simpler as

```js
  const args = await new Promise(resolve => worker.onerror = (...args) => resolve(args));
```

and then `args` will be an array, which avoids the awkward a/b/c properties and lets the `checkArguments` function do something like `assert_equals(args.length, 1)`.

File: workers/modules/shared-worker-parse-error-failure.html [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    };
+    worker.onerror = () => reject(false);
+  });
+  assert_precondition(
+    supportsModuleWorkers,
+    "Static import must be supported on module shared worker to run this test."
+  );
+});
+
+const checkArguments = (args) => {
```
Comment (domenic) [inline workers/modules/shared-worker-parse-error-failure.html]: To deduplicate this you may want to move it to a file in `workers/resources`, change it to `window.checkErrorArguments = ...`, and then use e.g. `<script src="resources/check-error-arguments.js"></script>` to get the function in both files.

**Review by domenic — CHANGES_REQUESTED** [review]

File: workers/modules/shared-worker-parse-error-failure.html (hunk 7 above)
Comment (elkurin) [reply workers/modules/shared-worker-parse-error-failure.html]: I created workers/support/check-error-arguments.js.

File: workers/modules/shared-worker-parse-error-failure.html [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+// the top-level script from the parse error invoked by the statically imported
+// script which is the one we want to check in this test.
+promise_setup(async () => {
+  const scriptURL = 'resources/static-import-worker.js';
+  const worker = new SharedWorker(scriptURL, { type: 'module' });
+  const supportsModuleWorkers = await new Promise((resolve, reject) => {
+    worker.port.onmessage = e => {
+      resolve(e.data.length == 1 && e.data[0] == 'export-on-load-script.js');
+    };
+    worker.onerror = () => reject(false);
```
Comment (domenic) [inline workers/modules/shared-worker-parse-error-failure.html]:
```suggestion
    worker.onerror = () => resolve(false);
```

File: workers/support/check-error-arguments.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+window.checkErrorArguments = args => {
+  assert_equals(args.length, 1);
+  assert_equals(args[0].constructor, Event);
+}
```
Comment (domenic) [inline workers/support/check-error-arguments.js]:
```suggestion
};
```

**Review by domenic — APPROVED** [review]
LGTM with nits. Thank you for your patience!

**Comment by elkurin** [comment]
Thanks so much @domenic for a lot of comments!
